docs/internals/optimizer.rst

```diff
@@ -0,0 +1,1254 @@
+
+The ControlFlowSimplifier does record the presence or absence of ``break``
+and ``continue`` statements during its traversal.
+
+Prerequisite: Disambiguator, FunctionHoister, ForLoopInitRewriter.
+Important: Introduces EVM opcodes and thus can only be used on EVM code for now.
+
+.. _dead-code-eliminator:
+
+DeadCodeEliminator
+^^^^^^^^^^^^^^^^^^
+
+This optimization stage removes unreachable code.
+
+Unreachable code is any code within a block which is preceded by a
+leave, return, invalid, break, continue, selfdestruct or revert.
+
+Function definitions are retained as they might be called by earlier
+code and thus are considered reachable.
+
+Because variables declared in a for loop's init block have their scope extended to the loop body,
+we require ForLoopInitRewriter to run before this step.
+
+Prerequisite: ForLoopInitRewriter, Function Hoister, Function Grouper
+
+
+.. _expression-inliner:
+
+ExpressionInliner
+^^^^^^^^^^^^^^^^^
+
+This optimizer component modifies an AST in place, inlining functions that can be
+inlined inside functional expressions, i.e. functions that:
+
+ - return a single value.
+ - have a body like r := <functional expression>.
+ - neither reference themselves nor r in the right hand side.
+
+Furthermore, for all parameters, all of the following need to be true:
+
+ - The argument is movable.
+ - The parameter is either referenced less than twice in the function body, or the argument is rather cheap ("cost" of at most 1 like a constant up to 0xff).
+
+This component can only be used on sources with unique names.
+
+.. _for-loop-condition-out-of-body:
+
+ForLoopConditionOutOfBody
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Reverses the transformation of ForLoopConditionIntoBody.
+
+For any movable ``c``, it turns
+
+::
+
+    for { ... } 1 { ... } {
+    if iszero(c) { break }
+    ...
+    }
+
+into
+
+::
+
+    for { ... } c { ... } {
+    ...
+    }
+
+and it turns
+
+::
+
+    for { ... } 1 { ... } {
+    if c { break }
+    ...
+    }
+
+into
+
+::
+
+    for { ... } iszero(c) { ... } {
+    ...
+    }
+
+The LiteralRematerialiser should be run before this step.
+
+.. _function-specializer:
+
+FunctionSpecializer
+^^^^^^^^^^^^^^^^^^^
+
+This step specializes the function with its literal arguments.
+
+If a function, say, ``function f(a, b) { sstore (a, b)}``, is called with literal arguments, for
+example, ``f(x, 5)``, where ``x`` is an identifier, it could be specialized by creating a new
+function ``f_1`` that takes only one argument, i.e.,
+
+::
+
+    function f_1(a_1) {
+    let b_1 := 5
+    sstore(a_1, b_1)
+    }
+
+Other optimization steps will be able to make more simplifications to the function. The
+optimization step is mainly useful for functions that would not be inlined.
+
+Prerequisites: Disambiguator, FunctionHoister
+
+LiteralRematerialiser is recommended as a prerequisite, even though it's not required for
+correctness.
+
+.. _literal-rematerialiser:
+
+LiteralRematerialiser
```

Comment:
Please move into "expression scale"

Reply:
Done, I couldn't find any documentation on the literal rematerialiser yet. Did I miss it somewhere?

Reply:
I couldn't find anything either.